internal/pkg/csumwriter/csumwriter.go

```diff
@@ -0,0 +1,91 @@
+package csumwriter
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+
+	"github.com/cespare/xxhash/v2"
+)
+
+const (
+	ChecksumLen = 8
+)
+
+type ChecksumWriter struct {
+	dataFile     io.Writer
+	checksumFile io.Writer
+	chunkSize    int
+	buf          []byte
+}
+
+func NewChecksumWriter(dataFile, checksumFile io.Writer, chunkSize int) *ChecksumWriter {
+	return &ChecksumWriter{
+		dataFile:     dataFile,
+		checksumFile: checksumFile,
+		chunkSize:    chunkSize,
+		buf:          make([]byte, 0, chunkSize),
+	}
+}
+
+func (cw *ChecksumWriter) Write(p []byte) (int, error) {
+	totalWritten := 0
+
+	for len(p) > 0 {
+		n := cw.chunkSize - len(cw.buf)
+		toWrite := len(p)
+		if toWrite > n {
+			toWrite = n
+		}
+
+		cw.buf = append(cw.buf, p[:toWrite]...)
+		p = p[toWrite:]
+		totalWritten += toWrite
+
+		if len(cw.buf) == cw.chunkSize {
+			if err := cw.flushChunk(); err != nil {
+				return totalWritten, err
+			}
+		}
+	}
+
+	return totalWritten, nil
+}
+
+func (cw *ChecksumWriter) flushChunk() error {
+	if len(cw.buf) == 0 {
+		return nil
+	}
+
+	n, err := cw.dataFile.Write(cw.buf)
+	if err != nil {
+		return fmt.Errorf("failed to write data chunk: %w", err)
+	}
+	if n != len(cw.buf) {
+		return fmt.Errorf("short write to data file: wrote %d, expected %d", n, len(cw.buf))
+	}
```

Comment:
This error check isn't necessary because `Write` of `io.Writer` returns an error if it couldn't write the whole data. cf. https://pkg.go.dev/io#Writer